[WIP] Support for Azure Batch - Pool

jcorioland opened the pull request:
This PR will brings support for Azure Batch Pool.
Still work in progress. 

Will fix #2427 
Depends on #2428 

Azure Batch Pool:

- [ ] Data Source
- [ ] Resource
- [ ] Tests
- [x] Examples
- [ ] Website documentation

tombuildsstuff commented:
@jcorioland thanks for splitting this out :)

Given this is dependent on #2427 - just to let you know I'm going to take a look through [the commit specific to this PR](https://github.com/terraform-providers/terraform-provider-azurerm/pull/2461/commits/8e754ee3d083db048f883e3f13bf37bcc6b6f92c), rather than attempting to re-review the bits present in the other PR (to hopefully avoids any conflicts)

tombuildsstuff commented:
@jcorioland I've merged master into this branch so those changes don't appear in the diff now that the other PR's been merged, I hope you don't mind

jcorioland commented:
@tombuildsstuff nope, this is perfect, thank you.
Was looking on the issue #2458 - not sure it require a new resource and being implemented in another PR as the start task is actually part of the [Batch PoolProperties in the Azure SDK for Go](https://github.com/Azure/azure-sdk-for-go/blob/199e71492c8f8ffb9d1d9432d425a1c8731fba19/services/batch/mgmt/2017-09-01/batch/models.go#L2226).
So I just need to add required properties in the schema of resource/data source and implement the start task in the Create/Update/Read functions.

What do you think about? Better to do it in this PR or a new one?
Thanks

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+			},
+			"storage_image_reference": {
+				Type:     schema.TypeSet,
+				Optional: true,
```
I believe we can remove this as it's a datasource?
```suggestion
```

jcorioland replied:
OK.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+				Type:     schema.TypeSet,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
```
I believe we can remove this as it's a datasource?
```suggestion
```

jcorioland replied:
OK.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				MaxItems: 1,
```
I believe we can remove this as it's a datasource?
```suggestion
```

jcorioland replied:
OK.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Should the ForceNew be removed and optional changed to Computed?

jcorioland replied:
Yep, it should. Done.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+
+						"offer": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Should the ForceNew be removed and optional changed to Computed?

jcorioland replied:
Yep it should. Done.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+
+						"publisher": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Should the ForceNew be removed and optional changed to Computed?

jcorioland replied:
Yep it should. Done.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+
+						"version": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Should the ForceNew and optional be removed?

jcorioland replied:
Yep it should. Done.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+	d.Set("vm_size", resp.VMSize)
+
+	if resp.ScaleSettings != nil {
+		if resp.ScaleSettings.AutoScale != nil {
```
Should these properties potentially be moved to two separate `auto_scale` and `fixed_scape` blocks?

jcorioland replied:
I added dedicated blocks for auto and fixed scale, it make more sense.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+
+		imageReference := resp.DeploymentConfiguration.VirtualMachineConfiguration.ImageReference
+
+		if err := d.Set("storage_image_reference", schema.NewSet(resourceArmVirtualMachineStorageImageReferenceHash, flattenAzureRmBatchPoolImageReference(imageReference))); err != nil {
```
Could we have `flattenAzureRmBatchPoolImageReference()` return a set rather then creating it here?

jcorioland replied:
Done.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+		imageReference := resp.DeploymentConfiguration.VirtualMachineConfiguration.ImageReference
+
+		if err := d.Set("storage_image_reference", schema.NewSet(resourceArmVirtualMachineStorageImageReferenceHash, flattenAzureRmBatchPoolImageReference(imageReference))); err != nil {
+			return fmt.Errorf("[DEBUG] Error setting AzureRM Batch Pool Image Reference: %#v", err)
```
I think we can remove the debug here
```suggestion
			return fmt.Errorf("Error setting AzureRM Batch Pool Image Reference: %#v", err)
```

jcorioland replied:
Ok.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+		}
+	}
+
+	if resp.DeploymentConfiguration != nil &&
```
This could be simpler if we:
```go
vmCfg := resp.DeploymentConfiguration.VirtualMachineConfiguration

if err := d.Set("storage_image_reference", flattenAzureRmBatchPoolImageReference(vmCfg.imageReference)); err != nil {
```

jcorioland replied:
Ok.

katbyte commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,159 @@
+			return fmt.Errorf("[DEBUG] Error setting AzureRM Batch Pool Image Reference: %#v", err)
+		}
+
+		d.Set("node_agent_sku_id", resp.DeploymentConfiguration.VirtualMachineConfiguration.NodeAgentSkuID)
```
This could then become: `d.Set("node_agent_sku_id", vmCfg.NodeAgentSkuID)`

jcorioland replied:
Ok.

katbyte commented on azurerm/data_source_batch_pool_test.go:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	data "azurerm_batch_pool" "test" {
+		name 							  = "${azurerm_batch_pool.test.name}"
```
Could we fix the alignment here?

jcorioland replied:
Done.

katbyte commented on azurerm/data_source_batch_pool_test.go:
```diff
@@ -0,0 +1,89 @@
+		node_agent_sku_id	= "batch.node.ubuntu 16.04"
+	
+		storage_image_reference {
+					publisher = "Canonical"
```
Could we fix the alignment here?

jcorioland replied:
Done.

katbyte commented on azurerm/helpers/azure/batch_pool.go:
```diff
@@ -0,0 +1,8 @@
+package azure
+
+type BatchPoolScaleMode string
```
We don't usually do this for missing constants, any thoughts @tombuildsstuff ?

jcorioland replied:
Actually, since I added dedicated blocks for fixed and auto scale these constants (and the scale_mode field) are not required anymore. I test directly on one of the block to be defined to determine what mode should be used for scaling.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,476 @@
+
+	scaleMode := azure.BatchPoolScaleMode(d.Get("scale_mode").(string))
+
+	if scaleMode == azure.BatchPoolAutoScale {
```
I think it would be best to put these properties into a `fixed_scale` and `auto_scale` blocks that conflict with each other, and they we can drop the comparison here.

jcorioland replied:
Done.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,476 @@
+		autoScaleEvaluationInterval := d.Get("autoscale_evaluation_interval").(string)
+		autoScaleFormula := d.Get("autoscale_formula").(string)
+
+		if autoScaleFormula == "" {
```
If this was in a sub block, we could make it required and have it validated during plan

jcorioland replied:
Done.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,476 @@
+		return fmt.Errorf("Error: storage image reference should be defined")
+	}
+
+	storageImageRef := storageImageReferenceSet.List()[0].(map[string]interface{})
```
Could we separate this out into an `expand` function?

jcorioland replied:
Done.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,476 @@
+		return fmt.Errorf("Error: scale mode should be either AutoScale of FixedScale")
+	}
+
+	if _, err = client.Update(ctx, resourceGroup, accountName, poolName, parameters, ""); err != nil {
```
It looks like the majority of the create and update function is the same? if this is the case maybe we could merge them, thoughts?

jcorioland replied:
Not exactly, but I have done some refactoring to avoid duplicate code. Should be better now, tell me.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,476 @@
+	return resourceArmBatchPoolRead(d, meta)
+}
+
+func resourceArmBatchPoolDelete(d *schema.ResourceData, meta interface{}) error {
```
Could we put this under the read function?

jcorioland replied:
Done.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,476 @@
+	d.Set("vm_size", resp.VMSize)
+
+	if resp.ScaleSettings != nil {
+		if resp.ScaleSettings.AutoScale != nil {
```
As above, i think these should be in two separate blocks

jcorioland replied:
Done.

katbyte commented on azurerm/resource_arm_batch_pool_test.go:
```diff
@@ -0,0 +1,319 @@
+	vm_size				   = "Standard_A1"
+	scale_mode			   = "Fixed"
+	target_dedicated_nodes = 2
+	node_agent_sku_id	= "batch.node.ubuntu 16.04"
```
Could we fix the alignment here?

jcorioland replied:
Done.

katbyte commented on azurerm/resource_arm_batch_pool_test.go:
```diff
@@ -0,0 +1,319 @@
+	scale_mode			   		  = "Auto"
+	autoscale_evaluation_interval = "PT15M"
+	autoscale_formula			  = <<EOF
+	startingNumberOfVMs = 1;
```
Could we fix the assignment alignment here?

jcorioland replied:
Done.

katbyte commented on examples/batch/main.tf:
```diff
@@ -30,4 +30,48 @@ resource "azurerm_batch_account" "batch" {
+  resource_group_name    = "${azurerm_resource_group.rg.name}"
+  account_name 		       = "${azurerm_batch_account.batch.name}"
+  display_name		       = "Fixed Scale Pool"
+  vm_size				         = "Standard_A1"
```
Could we run terraform ft on this file to fix the alignment? you can just point it at the folder and it'll find and fmt all terraform files 🙂

jcorioland replied:
Done.

katbyte commented on website/docs/r/batch_pool.html.markdown:
```diff
@@ -0,0 +1,120 @@
+}
+
+resource "azurerm_batch_pool" "test" {
+	name                   		    = "testaccpool"
```
Could we fix the alignment here?

jcorioland replied:
Done.

katbyte commented on website/docs/r/batch_pool.html.markdown:
```diff
@@ -0,0 +1,120 @@
+
+* `autoscale_formula` - The formula used to autoscale the pool when using scale mode `Auto`.
+
+* `storage_image_reference` - The reference of the storage image used by the nodes in the Batch pool.
```
Could we add an import section here?

jcorioland replied:
Done.

jcorioland commented:
@katbyte - Thanks for the review. I've done some refactoring and most of the changes you've asked for are done.
I will implement start task part of this PR, it does not make sense to create another one, indeed.

jcorioland commented on azurerm/helpers/azure/batch_pool.go:
```diff
@@ -0,0 +1,468 @@
+	}
+
+	return schema.NewSet(resourceArmBatchPoolImageReferenceHash, []interface{}{result})
+}
```
I've done a bit different than what you've proposed using make() to create an empty list and return it if image is nil, but I think it's the same.

jcorioland commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,329 @@
+		stopPendingResizeOperation := d.Get("stop_pending_resize_operation").(bool)
+		if !stopPendingResizeOperation {
+			return fmt.Errorf("Error updating the Batch pool %q (Resource Group %q) because of pending resize operation. Set flag `stop_pending_resize_operation` to true to force update", poolName, resourceGroup)
+		}
```
The resize operation occurs as soon as you create or update the pool. I understand from your remark that I should wait for it after a create/update. But in the case of an update, the pool can be in a pending state (if the user update something through the portal, for example).
I think we can still expose this field to offer the choice to wait or not. What do you think about?

jcorioland replied:
I let the option to force to stop any pending resize operation when doing an update, but I've also refactor my code to make sure both Create and Update operations wait for the Pool to be ready before returning. I think we're all covered with that.

jcorioland commented on examples/batch/main.tf:
```diff
@@ -1,10 +1,6 @@
-  # client_id       = "..." 
-  # client_secret   = "..."
-  # tenant_id       = "..."
+  # if you're using a Service Principal (shared account) then either set the environment variables, or fill these in:   # subscription_id = "..."  # client_id       = "..."   # client_secret   = "..."  # tenant_id       = "..."
```
for the record, `terraform fmt` inline the comments :-)

tombuildsstuff commented on azurerm/data_source_batch_pool.go:
```diff
@@ -0,0 +1,92 @@
+	d.Set("name", name)
+	d.Set("account_name", accountName)
+	d.Set("resource_group_name", resourceGroup)
+	d.Set("vm_size", resp.VMSize)
```
The Azure API returns the representation of a resource at the last time it was modified; meaning if it’s not been modified in a while (for example if it was made during a Preview API) the structure returned from the API may not necessarily match the result we expect. As such we ensure we access these fields in the Properties block, since that’s where we expect them

katbyte commented:
@jcorioland,

It appears to be a casing issue with `vm_size` being returned in all caps from the API? causing it to want to re-create the resource. Looks like that property is going to have to be case insensitive (and a bug should probably be opened with Azure as that seems like an error)

jcorioland commented:
@katbyte thanks a lot, making the `vmSize` case insensitive solved the issue. All tests pass now!
I think that I've fixed all the issues that you and @tombuildsstuff raised out during your reviews.
LMK if I need to do other changes.
Thanks

tombuildsstuff commented:
Data Source test passes:

<img width="490" alt="screenshot 2018-12-23 at 15 15 48" src="https://user-images.githubusercontent.com/666005/50385028-a3ec3100-06c5-11e9-8dde-e2631172fcf9.png">

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,518 @@
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  1,
+						},
```
Could we get some validation here?

jcorioland replied:
OK.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,518 @@
+						"target_low_priority_nodes": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  0,
```
Could we get some validation here?

jcorioland replied:
OK.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,518 @@
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
```
Could we validate the ID: `azure.ValidateResourceID`

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,518 @@
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
```
Could we get at least a `validate.NoEmptyStrings` here?

jcorioland replied:
OK.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,518 @@
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
```
Could we get at least a `validate.NoEmptyStrings` here?

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,518 @@
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
```
Could we get at least a `validate.NoEmptyStrings` here?

jcorioland replied:
OK.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,518 @@
+							Required:         true,
+							ForceNew:         true,
+							DiffSuppressFunc: suppress.CaseDifference,
+						},
```
Could we get at least a `validate.NoEmptyStrings` here?

jcorioland replied:
OK.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,518 @@
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
```
Could we get at least a `validate.NoEmptyStrings` here?

jcorioland replied:
OK.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,518 @@
+						"command_line": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
```
Could we get at least a `validate.NoEmptyStrings` here?

jcorioland replied:
OK.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -0,0 +1,518 @@
+
+						"user_identity": {
+							Type:     schema.TypeList,
+							Required: true,
```
Should this we required when all the sub properties are optional?

jcorioland replied:
I was not sure for this one: user identity is mandatory, but you should specify either a user name or an auto identity. This is why I put the `user_identity` block mandatory but the two properties inside optional. What the best way to handle that? Adding some extra validation to the `user_identity` block to check that at least one type of identity has been specified?

katbyte replied:
If one of those values is required i would add an additional check & document it in the create/update functions for now.

jcorioland commented:
Hi @katbyte @tombuildsstuff - Happy New Year! 
Thanks for your reviews.
 
I've added some validations, except for the `user_identity`, not sure on the best way to handle that case? (cf. my comment). 
I don't have any issue with the acceptance test on autoscale, but I've added a suppress func to trim the string, as you've suggested.

Thank you

katbyte commented:
Happy new year @jcorioland! Thanks for the commits.

If one of those values is required i would add a check for it into the create/update functions. A `CutomizeDiff` would be better to check during the plan but it doesn't work with interpolated values at the moment.

jcorioland commented:
@katbyte thanks! I've just added validation for user identity. Should be good now.
